### ISSUE TITLE
multiple instances of RtAudio accessing the same CoreAudio device

### ISSUE THREAD

**Issue by rick-spada** [thestk/rtaudio#205]
I'm puzzled by the comment in the RtAudio API note:

> It is not possible to have multiple instances of RtAudio accessing the same CoreAudio device.

I would think that since `RtApiCore::probeDeviceOpen` is calling `AudioDeviceCreateIOProcID` that it would be possible since the documentation for that clearly states:

> AudioDeviceIOProcIDs allow for the client to register the same function pointer with a device multiple times

-- Rick

**Comment by garyscavone** [comment]
Perhaps that statement was wrong. Have you tried it?

That said, I don’t see why someone would want to have multiple instances of RtAudio accessing the same device.

**Comment by rick-spada** [comment]
Yes, I tried it and it doesn't work.  I only get output on the last channel I try to open.

As for wanting to have multiple instances of RtAudio accessing the same device, I'm trying to access each channel (left or right) independently vs having to access them as a pair.  It will greatly simplify some code on our end if we could.  FWIW, I can do this on Windows with WASAPI.

**Comment by rick-spada** [comment]
OK, I spent some more time looking at the code and I got this working.  I believe there's a bug in `RtAPpiCore::startStream`.  The lines of code that start the audio device are always passing in the `callbackHandler`, not the proc ID that was given in `probeDeviceOpen`.

**Comment by rick-spada** [comment]
I have a fix for how RtApiCore uses `procId`'s and would like to create a pull request so as to get it into RtAudio.  Unfortunately, I do not have permissions to do so.  I've attached a diff of RtAudio.cpp.
[RtAudio.cpp.diff.txt](https://github.com/thestk/rtaudio/files/3345413/RtAudio.cpp.diff.txt)

**Comment by garyscavone** [comment]
You should not need permission to submit a PR. Can you try again?

**Comment by rick-spada** [comment]
Yup, tried again and got the same error:

`remote: Permission to thestk/rtaudio.git denied to rick-spada.`

I'm trying to create the pull request using the following:

`git push origin fix_proc_id_usage_core_audio`

Where `fix_proc_id_usage_core_audio` is the name of my local branch.

**Comment by almogh52** [comment]
ETA on when this is going to be fixed?

**Comment by rick-spada** [comment]
Dunno...I can't seem to create a pull request so there's no way for me to move the changes forward.

**Comment by theicfire** [comment]
You mentioned " I don’t see why someone would want to have multiple instances of RtAudio accessing the same device.", but isn't simply wanting to connect to the microphone and speaker of the same device an example of this?

I have this precise problem right now, and I'm finding that trying to connect two RtAudio instances to the same device doesn't work. It took me awhile to notice this issue because there's no error. Just one of the callbacks won't run.

What's the recommended way of doing this? I could make a single RtAudio instance, but now the code is quite confusing:
1) If I have a microphone/speaker that are are different devices, make two RtAudio instances
2) If they're on the same device, make a single instance

Is there a better way?

**Comment by theicfire** [comment]
Oh, maybe the answer is simply that I'm on 5.1.0 and I should move to 5.2.0. I guess the documentation needs to be updated as well?

**Comment by theicfire** [comment]
Hmm, I'm getting the same behavior in both versions: The second rtaudio instance that connects doesn't have callbacks run. The first will.

**Comment by garyscavone** [comment]
You can make use of different devices for input and output through a singe instance of RtAudio (you don't need different instances).

**Comment by theicfire** [comment]
Oh, of course! For my use case, two instances seems like a nicer API to work with, so I'll stick with that. Thanks for the response!
